[UnifiedPDF] "Look Up ..." context menu item should perform a dictionary lookup of the current selection

aprotyas opened the pull request:
#### f8d35c16351cd214169214955c2507ba338a9a5e
<pre>
[UnifiedPDF] &quot;Look Up ...&quot; context menu item should perform a dictionary lookup of the current selection
<a href="https://bugs.webkit.org/show_bug.cgi?id=269434">https://bugs.webkit.org/show_bug.cgi?id=269434</a>
<a href="https://rdar.apple.com/122991260">rdar://122991260</a>

Reviewed by Tim Horton.

We introduced the &quot;Look Up ...&quot; context menu item in 274611@main, but
had deferred actually performing selection look up for a later commit.

This patch implements said functionality in
UnifiedPDFPlugin::searchInDictionary and adds the plumbing for the
relevant context menu item to call into it.

Moreover, this commit implements the performDictionaryLookupAtLocation
interface, which allows other lookup paths to work too, namely for the
lookup gesture recognizers in WKContentView and for -[WKWebView
quickLookWithEvent].

* Source/WebCore/editing/cocoa/DictionaryLookup.mm:
(WebCore::DictionaryLookup::stringForPDFSelection):
* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
(WebKit::UnifiedPDFPlugin::performDictionaryLookupAtLocation):
(WebKit::UnifiedPDFPlugin::searchInDictionary):
(WebKit::UnifiedPDFPlugin::selectionBoundsForFirstPageInDocumentSpace const):
(WebKit::UnifiedPDFPlugin::showDefinitionForAttributedString):

Canonical link: <a href="https://commits.webkit.org/274770@main">https://commits.webkit.org/274770@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02281d350cb33100ef98155f5a52385622e13110

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39946 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18957 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42321 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42490 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42253 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21864 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16287 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33278 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40520 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15975 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34529 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43769 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36263 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16380 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8961 "Build was cancelled. Recent messages:OS: Monterey (12.6.8), Xcode: 13.4.1; Checked out pull request; Reviewed by Tim Horton; Compiled WebKit (warnings); 16 flakes 37 missing results; Uploaded test results; layout-tests (exception)") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5272 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2410,11 +2411,67 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
+    return std::nullopt;
+}
+
+bool UnifiedPDFPlugin::showDefinitionForAttributedString(RetainPtr<NSAttributedString>&& string, const IntRect& rectInDocumentSpace)
```
I just realized that `PDFPlugin` has a function with a very similar signature. I should unify (ha!) them somehow, or at least delegate the "inform webpage that a dictionary lookup was performed" bit to `PDFPluginBase`.

aprotyas replied:
I'll do this in a future commit. Landing this as-is for now.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2410,11 +2411,67 @@ static IntRect computeMarqueeSelectionRect(const WebCore::IntPoint& point1, cons
+    TextIndicatorData dataForSelection;
+    dataForSelection.selectionRectInRootViewCoordinates = rectInRootView;
+    dataForSelection.textBoundingRectInRootViewCoordinates = rectInRootView;
+    dataForSelection.contentImageScaleFactor = m_scaleFactor * m_documentLayout.scale();
```
Weird to set this when you don't have a contentImage but maybe it gets used for something else?

aprotyas replied:
Yes, this is not necessary. I was following precedence set by `PDFPlugin::showDefinitionForAttributedString` and `WebPage::dictionaryPopupInfoForSelectionInPDFPlugin`. I'll drop it for now with the assumption that it's redundant.

aprotyas committed:
[UnifiedPDF] "Look Up ..." context menu item should perform a dictionary lookup of the current selection
https://bugs.webkit.org/show_bug.cgi?id=269434
rdar://122991260

Reviewed by Tim Horton.

We introduced the "Look Up ..." context menu item in 274611@main, but
had deferred actually performing selection look up for a later commit.

This patch implements said functionality in
UnifiedPDFPlugin::searchInDictionary and adds the plumbing for the
relevant context menu item to call into it.

Moreover, this commit implements the performDictionaryLookupAtLocation
interface, which allows other lookup paths to work too, namely for the
lookup gesture recognizers in WKContentView and for -[WKWebView
quickLookWithEvent].

* Source/WebCore/editing/cocoa/DictionaryLookup.mm:
(WebCore::DictionaryLookup::stringForPDFSelection):
* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::performContextMenuAction):
(WebKit::UnifiedPDFPlugin::performDictionaryLookupAtLocation):
(WebKit::UnifiedPDFPlugin::searchInDictionary):
(WebKit::UnifiedPDFPlugin::selectionBoundsForFirstPageInDocumentSpace const):
(WebKit::UnifiedPDFPlugin::showDefinitionForAttributedString):

Canonical link: https://commits.webkit.org/274770@main
